### ISSUE TITLE
How to make the directional wind stress reduction nodal attribute

### ISSUE THREAD

**Issue by krober10nd** [zcobell/QADCModules#32]
* Firstly, thanks to @zcobell for helping me through this process and writing good, open source code. 
* This short write-up is for others who may be wanting to create the directional wind stress reduction nodal attribute that's used inside ADCIRC to take into account roughness elements when forcing your model with certain atmospheric models.
* I realize this is not an "issue" per se but due to the lack of other communication channels I hope this suffices. This could be modified for the wiki site perhaps. 

Okay...
* First, you'll need to install ADCIRCmodules. Using CMake this is a breeze. The only difficulty I had was finding OpenMP, which can be used to accelerate this computationally expensive process to calculate the directional wind stress nodal attribute. 
* Then you'll need to download the CCAP landcover usage dataset available [here](https://coast.noaa.gov/digitalcoast/data/ccapregional.html). I downloaded the 2016 tif file from there. 
* After this, I put together a script that looks something like: 
```
 #include <iostream>
 #include <memory>
 #include "adcircmodules.h"
 
 int main() {
   using namespace Adcirc::Geometry;
   using namespace Adcirc::Interpolation;
   using namespace Adcirc::ModelParameters;
 
   std::unique_ptr<Mesh> m(new Mesh("../binera25m_wFP_tide_only.14"));
   m->read();
   m->defineProjection(4326, true);
 
   std::unique_ptr<NodalAttributes> fort13(new NodalAttributes("../binera25m_wFP_tide_only.13", m.get()));
   fort13->read();
   auto midx2 = fort13->locateAttribute("surface_directional_effective_roughness_length");
 
   std::unique_ptr<Griddata> g(new Griddata(m.get(), "consu_epsg26918.tif", 26918));
 
   for (size_t i = 0; i < m->numNodes(); ++i) {
       //if( i < 50){
         g->setInterpolationFlag(i, Adcirc::Interpolation::Average);
       //}
       //else{
       //  g->setInterpolationFlag(i, Adcirc::Interpolation::NoMethod);
       //}
   }
 
   g->setShowProgressBar(true);
   g->readLookupTable("../CCAP_DWind.table");
   g->setRasterInMemory(false);
   std::vector<std::vector<double>> r = g->computeDirectionalWindReduction(true);
 
   for (size_t i =0; i < m->numNodes(); ++i){
       //if(i < 50){
       //  std::cout<< r[i][0] <<std::endl;
       //}
     fort13->attribute(midx2,i)->setValue(r[i]);
   }
 
   fort13->write("../binera25m_wFP_w_dir_ws.13");
 }
```
* Note that you'll have to create a dummy fort.13  with the directional wind stress nodal attribute or add on the directional wind stress attribute to an existing fort.13 file. 
Basically, if you're creating a fort.13 from scratch you'll need to have the following in your fort.13 file. 
```
surface_directional_effective_roughness_length 
m
12
0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0
surface_directional_effective_roughness_length
0
```
* If you're modifying a fort.13 to support this attribute, then you'll have to add the default values to the top of the file and the user defined values to an appropriate part in the fort.13 file. 

* Compile your script (I called it `make_dwind.cpp`) by linking to the location where the ADCIRCmodule header files are located and link to the shared library. 
```
g++ -I /usr/local/include/adcircmodules/  -ladcircmodules -std=c++14 make_dwind.cpp -o make_dwind.x
```
* It's important at this point, to stress a couple things.
1. You'll need a look up table (LUT) called something like `CCAP_DWind.table` which one could save the text below into a file. These values are used to assign roughness values at the nodes so they're important to get right. Place this LUT file in your working directory and make sure the path is correct in your script to-be-compiled.
```
   2    0.500     :High Intensity Developed
  3    0.390     :Medium Intensity Developed
  4    0.500     :Low Intesnity Developed
  5    0.330     :Developed Open Space
  6    0.060     :Cultivated Land
  7    0.060     :Pasture/Hay
  8    0.040     :Grassland
  9    0.650     :Deciduous Forest
 10    0.720     :Evergreen Forest
 11    0.710     :Mixed Forest
 12    0.120     :Scrub/Shrub
 13    0.550     :Palustrine Forested Wetland
 14    0.110     :Palustrine Scrub/Shrub Wetland
 15    0.110     :Palustrine Emergent Wetland
 16    0.550     :Estuarine Forested Wetland
 17    0.120     :Estuarine Scrub/Schrub Wetland
 18    0.110     :Estuarine Emergent Wetland
 19    0.090     :Unconsolidated Shore
 20    0.090     :Bare Land
 21    0.001     :Open Water
 22    0.040     :Palustrine Aquatic Bed
 23    0.040     :Estuarine Aquatic Bed
*The class has default value(=-9999) will be skiped in mapping
```

2. Secondly, test the program at first with only a handful of nodes (say 50 or so) by uncommenting the code where I assign the interpolation flag "NoMethod". This way you won't have to wait hours to see a code syntax error appear.
3. Reproject the CCAP tif into a ESPG projection as it has to be in a UTM coordinate system. I chose ESPG:26918 for the NY/NJ area but this depends on where your model is. I accomplished this re-projection of the tif file using gdal's `gdalwarp` method. Such as: 
```
gdalwarp -t_srs ESPG:21918 source.tif output.tif
```
After you compile and run the code, you'll see a beautiful progress bar that will give you an estimated time for the process until completion. You can check and visualize this fort13 using [OceanMesh2D](https://github.com/CHLNDDEV/OceanMesh2D) in MATLAB by first reading the files in. 
```
m = msh('fname','fort.14','aux',{'myfort.13'}); 
plot(m,'type','dir','proj','lamb'); 
```
Note that by default it plots the infinity norm at each point (the largest directional wind stress reduction at each node of the grid). 

which yields something like this:

![dirws](https://user-images.githubusercontent.com/18619644/104629276-0e3ee800-5678-11eb-8df4-024b37fd9f7b.png)

* A close up
![dirws0](https://user-images.githubusercontent.com/18619644/104629296-139c3280-5678-11eb-8a4f-cd40763e96eb.png)

**Comment by zcobell** [comment]
I've added this to the [Doxygen documentation](http://zcobell.github.io/ADCIRCModules/directional_wind_reduction_8cpp-example.html).

**Comment by zcobell** [comment]
Resolved with #33
